### ISSUE TITLE
Support `@NonNull` annotation on local variables

### ISSUE THREAD

**Issue by msridhar** [uber/NullAway#561]
This would allow for more localized suppressions.  Say we have a method `foo(Object o)` in an annotated third-party library where `o` has no type annotation, but `o` really should be `@Nullable`, as `foo()` can handle being passed `null` as a parameter.  In code checked by NullAway, say you write:
```java
void bigMethod() {
  ...
  foo(null); // NullAway reports an error
  ...
}
```
I believe the only way to suppress the warning above currently is to write `@SuppressWarnings("NullAway")` on `bigMethod`, which could suppress many other valid warnings.  If we supported checking `@NonNull` annotations on locals, then you could do a more localized suppression:
```java
void bigMethod() {
  ...
  // suppress since we know foo can handle null
  @SuppressWarnings("NullAway")
  @NonNull Object p = null;
  foo(p); 
  ...
}
```
This works because Java supports writing `@SuppressWarnings` on local variable declarations.  Additionally, it's just kind of weird that you currently don't get a warning from NullAway for writing `@NonNull Object p = null;` right now.

**Comment by sophiastr** [comment]
Hello, i want to deal with this issue.
Do you still need help?
I am beginner, is that a problem?
I will work really hard in order to make my first contribution.

**Comment by msridhar** [comment]
@sophiastr we'd be happy to have your help on this issue.  You will need to dig through and learn some javac APIs in order to accomplish this task.  I'd start as follows:

1. Write a failing test case.  You can add a test at the bottom of this file: https://github.com/uber/NullAway/blob/c9d89bf99d724b1824919eb010322e0763ac28b0/nullaway/src/test/java/com/uber/nullaway/NullAwayCoreTests.java It should contain a statement like `@NonNull Object p = null`, with `p` being a local variable.  And the test should fail since there should be a comment `// Bug: Diagnostic contains: something` before that statement, but no error is reported.
2. Look at the `matchAssignment()` code here: https://github.com/uber/NullAway/blob/c9d89bf99d724b1824919eb010322e0763ac28b0/nullaway/src/main/java/com/uber/nullaway/NullAway.java#L454 Right now this only handles assignments to fields.  Extend it to also report an error for the case of step 1.

This is not the final solution to this problem, since we will need to handle other cases as well.  But if you accomplish the above, we will be on our way to a final solution I think.

**Comment by msridhar** [comment]
In light of the resolution of https://github.com/jspecify/jspecify/issues/251 I'm going to close this one, as I don't think we will support this.  See https://github.com/jspecify/jspecify/issues/251 for alternate suppression mechanisms.
